components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Cdm/src/Services/CdmMessageProcessor.cs

```diff
@@ -160,47 +159,82 @@ public class CdmMessageProcessor : ICdmClient {
+                        if (result == false && retry == false) {
+                            retry = true;
+                        }
+                        else {
+                            retry = false;
+                        }
+                    } while (retry);
+
+                    _logger.Information("Successfully processed CDM data {count}" +
+                        " MonitoredItemSample records.", record.Value.Count);
                 }
 
-                foreach(var dataSetsList in _dataSetsCache.Values) {
-                    if (dataSetsList.Count == 0 || dataSetsList[0] == null) {
+                foreach (var record in _dataSetsCache) {
+                    if (record.Value.Count == 0 ||record.Value[0] == null) {
                         _logger.Error("DataSet list is empty ...");
                         continue;
                     }
-                    if (!GetEntityData(dataSetsList[0], out var partitionLocation, out var partitionDelimitor)) {
-                        if (!CreateEntityData(dataSetsList[0], out partitionLocation, out partitionDelimitor)) {
-                            _logger.Error("Failed to create CDM Entity for {PublilsherId}/{DataSetWriterId}/{MetaDataVersion}).",
-                                dataSetsList[0].PublisherId, dataSetsList[0].DataSetWriterId, dataSetsList[0].MetaDataVersion);
+                    var retry = false;
+                    do {
+                        var partition = GetOrCreateEntityDataPartition(record.Key,
+                            record.Value[0], out var persist, retry);
+                        if (partition == null) {
+                            _logger.Error("Failed to create CDM Entity for " +
+                                "{PublisherId}/{DataSetWriterId}/{DataSetClassId}/{MetaDataVersion}).",
+                                record.Value[0].PublisherId, record.Value[0].DataSetWriterId,
+                                record.Value[0].DataSetClassId, record.Value[0].MetaDataVersion);
                             continue;
                         }
-                        performSave = true;
-                    }
-                    await _storage.WriteInCsvPartition<DataSetMessageModel>(
-                        partitionLocation, dataSetsList, partitionDelimitor);
+                        performSave |= persist;
+
+                        var csvTrait = partition.ExhibitsTraits.Item("is.partition.format.CSV");
+                        var partitionUrl = _cdmCorpus.Storage.CorpusPathToAdapterPath(partition.Location);
+                        var partitionDelimitor = csvTrait?.Arguments?.FetchValue("delimiter") ?? kCsvPartitionsDelimiter;
+                        var result = await _storage.WriteInCsvPartition<DataSetMessageModel>(
+                            partitionUrl, record.Value, partitionDelimitor);
+                        if (result == false && retry == false) {
+                            retry = true;
```

Comment:
this code is almost exactly a duplicate of the previous do while retry. Can you make it a helper function?